### PR TITLE
add unit tests for server selection

### PR DESCRIPTION
**Pull request by mbroadst**
## Description
This change factors out the `selectServers` method from `topology.js` such that we can begin to unit tests parts of server selection without needing to mock too much. Additionally, it removes code that attempts to connect a `Topology` if its not already connected - we can't provide this guarantee until we remove `connect` altogether. 

**Are there any files to ignore?**
There's a lot of files moving around. Everything inside `test/` can be ignored except for the `select_servers_tests.js` which are the newly introduced unit tests for selection.

### CODE REVIEW

File: test/tools/sdam_viz [hunk 1]
```diff
@@ -85,8 +88,33 @@ async function run() {
   });
 
   await client.connect();
+
+  if (argv.workload) {
+    for (;;) {
```
Comment (daprahamian) [inline test/tools/sdam_viz]: Neat. Could we not do:

```js
while (!workloadInterrupt) {
```

File: test/core/functional/max_staleness_tests.js [hunk 2]
```diff
@@ -9,8 +9,8 @@ const expect = require('chai').expect,
   MongoError = require('../../../lib/core/error').MongoError,
   ReadPreference = require('../../../lib/core/topologies/read_preference');
 
-const rsWithPrimaryPath = f('%s/../spec/max-staleness/ReplicaSetWithPrimary', __dirname);
-const rsWithoutPrimaryPath = f('%s/../spec/max-staleness/ReplicaSetNoPrimary', __dirname);
+const rsWithPrimaryPath = f('%s/../../spec/max-staleness/ReplicaSetWithPrimary', __dirname);
```
Comment (daprahamian) [inline test/core/functional/max_staleness_tests.js]: Since we're here, can we use a template?

File: lib/core/sdam/topology_description.js [hunk 3]
```diff
@@ -404,7 +396,4 @@ function checkHasPrimary(serverDescriptions) {
   return TopologyType.ReplicaSetNoPrimary;
 }
 
-module.exports = {
```
Comment (daprahamian) [inline lib/core/sdam/topology_description.js]: Can we just leave it as a named export?

File: test/unit/sdam/server_selection/select_servers_tests.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  });
+
+  afterEach(function() {
+    this.sinon.restore();
+  });
+
+  it('should error immediately if timeout exceeds start time', function(done) {
+    const topology = new Topology('invalid:27019');
+    const start = process.hrtime();
+    start[0] = start[0] - 500;
```
Comment (daprahamian) [inline test/unit/sdam/server_selection/select_servers_tests.js]: Are we expecting a 500s timeout?

File: test/tools/sdam_viz (hunk 1 above)
Comment (mbroadst) [reply test/tools/sdam_viz]: yep, changed!

File: test/unit/sdam/server_selection/select_servers_tests.js (hunk 4 above)
Comment (mbroadst) [reply test/unit/sdam/server_selection/select_servers_tests.js]: yes, see the call to `selectServers` below. The third parameter is the timeout

**Pull request merged**